### PR TITLE
feat(embodied): bootstrap-training overlap when possible

### PR DESCRIPTION
**Pull request by aasivas**
Recreated the PR as commit checks require branch names to satisfy some rules. See #942 for the original PR and comments.

Performance Improvement: Asynchronous Environment Bootstrapping and Training Overlap

### Description
~~This PR introduces a new `warmup` method to the `EnvWorker` class to enable asynchronous environment preparation. The `warmup` function is designed to be called prior to `interact`, allowing environment bootstrapping to occur in parallel with `run_training`.
We also add a new `-1` configuration option for environment settings to explicitly denote CPU-only physics and rendering, and the function `get_group_accelerators` in `embodied_runner.py` to check if there is resource overlap.~~
Updated the PR based on feedback and copied suggested changes. Main difference is that warmup is not a function but an async handle (it is not called warmup anymore) and class member holding the bootstrap data is freed after consumption (e.g. set to `None`). 
There is also a new configuration for `runner`, `overlap_env_bootstrap`. This parameter needs to be set to True in the yaml file:
```
runner:
     overlap_env_bootstrap: True
``` 

### Motivation and Context
Currently, environment bootstrap time accounts for a significant portion (5%–20%) of each global step. In hardware configurations where `EnvWorker` processes do not compete for resources with the training actor (e.g., CPU-based rendering vs. NPU/GPU-based training), this bootstrap latency can be masked by overlapping it with the training phase.

While the initial global step time remains unchanged, subsequent steps see significant improvement in throughput.
*Preliminary Results:*
In a CPU + NPU environment (benchmarked on `libero_10` with Pi0.5 and PPO, with changes for minimal configuration), we observed a reduction in per-global-step time from 720s to approximately 620–640s.

### How has this been tested?
Validation is ongoing on our internal clusters using standard RLinf examples. Specifically:
Execution Script: `run_embodiment.sh`
Configuration: `libero_10_ppo_openpi_pi05.yaml` (with some selected minor adjustments from [Section 2.4 Minimum Test Case](https://rlinf.readthedocs.io/en/latest/rst_source/examples/embodied/pi0.html)).

### Additional information (optional, e.g., figures and logs):
Initial performance figures are posted in [comment](https://github.com/RLinf/RLinf/pull/942#issuecomment-4145842761).
_Further logs and performance figures will be attached upon completion of the test run._

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
(We are doing our due diligence to ensure that our changes do not break existing functionality. However, we are marking this box in case we miss something. If we are confident that our changes are non-breaking, we will remove the tick.)

### Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/config.py [hunk 1]
```diff
@@ -875,6 +875,9 @@ def validate_embodied_cfg(cfg):
         weight_sync_interval = cfg.runner.get("weight_sync_interval", 1)
         assert weight_sync_interval > 0, "weight_sync_interval must be greater than 0"
         cfg.runner.weight_sync_interval = weight_sync_interval
+        cfg.runner.overlap_env_bootstrap = bool(
+            cfg.runner.get("overlap_env_bootstrap", True)
```
Comment (QuanluZhang) [inline rlinf/config.py]: please make the default value false

Comment (Zengwen2001) [inline rlinf/config.py]: I agree with that

**Comment by QuanluZhang** [comment]
@aasivas thanks for your contribution! As the overlap path is newly added and it provides users a new configuration option, please add focused tests for prefetched bootstrap consumption and duplicate-prefetch protection, plus docs/config comments explaining runner.overlap_env_bootstrap and resource assumptions.

**Review by Zengwen2001 — APPROVED** [review]

File: docs/source-zh/rst_source/tutorials/user/yaml.rst [hunk 2]
```diff
@@ -747,11 +747,18 @@ runner
...
+    overlap_env_bootstrap: False
 
 ``runner.only_eval``：只运行评估，不进行训练。  
 
 ``runner.max_prompt_length``：最大提示长度（token 数）。
 
+``runner.overlap_env_bootstrap``： 
+将环境的 bootstrap（重置/reset）过程与 Actor 训练过程重叠，以隐藏重置延迟。 
+当环境重置较慢时，此功能非常有用。 
+**注意：** 仅在 ``env.enable_offload`` 为 False 时生效。
```
Comment (andylin-hao) [inline docs/source-zh/rst_source/tutorials/user/yaml.rst]: env.train.enable_offload? Check config.py

File: rlinf/runners/embodied_runner.py [hunk 3]
```diff
@@ -71,6 +71,9 @@ def __init__(
         self.critic = critic
         self.reward = reward
         self.weight_sync_interval = self.cfg.runner.weight_sync_interval
+        self.overlap_env_bootstrap = bool(
+            self.cfg.runner.get("overlap_env_bootstrap", False)
+        )
```
Comment (Lin-xs) [inline rlinf/runners/embodied_runner.py:76]: It's unnecessary to convert to `bool` here because `OmegaConf` will automatically process the type.

File: rlinf/config.py [hunk 4]
```diff
@@ -875,6 +875,13 @@ def validate_embodied_cfg(cfg):
         weight_sync_interval = cfg.runner.get("weight_sync_interval", 1)
         assert weight_sync_interval > 0, "weight_sync_interval must be greater than 0"
         cfg.runner.weight_sync_interval = weight_sync_interval
+        # Overlap environment bootstrap (reset) with actor training to hide reset latency.
+        # This is enabled only when offload is disabled to avoid resource contention.
+        # Note: If EnvWorker and Actor share the same accelerator, this may increase GPU memory
+        # pressure during the overlap period.
+        cfg.runner.overlap_env_bootstrap = bool(
```
Comment (Lin-xs) [inline rlinf/config.py:882]: Same issue.

File: tests/unit_tests/test_overlap_env_bootstrap.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
+# Copyright 2025 The RLinf Authors.
```
Comment (Lin-xs) [inline tests/unit_tests/test_overlap_env_bootstrap.py]: 2025 -> 2026

**Review by andylin-hao — APPROVED** [review]

**Review by QuanluZhang — APPROVED** [review]

File: rlinf/runners/embodied_runner.py (hunk 3 above)
Comment (aasivas) [reply rlinf/runners/embodied_runner.py:76]: I will reject this, because it is RLInf project style to explicitly cast to bool. Same style code appears more than 50 times, closest example is in the same file (embodied_runner.py) lines 98-100.

File: rlinf/config.py (hunk 4 above)
Comment (aasivas) [reply rlinf/config.py:882]: I will reject this, because it is RLInf project style to explicitly cast to bool. Same style code appears more than 50 times, closest example is in the same file (config.py) line 998.

```
946 |    runner_only_eval = cfg.runner.get("only_eval", None)
...
997 |    with open_dict(cfg):
998 |        cfg.runner.only_eval = bool(runner_only_eval)
...
```

File: tests/unit_tests/test_overlap_env_bootstrap.py (hunk 5 above)
Comment (aasivas) [reply tests/unit_tests/test_overlap_env_bootstrap.py]: Accepted, good catch. I copy-pasted the copyright notice and forgot to update the year.

File: docs/source-zh/rst_source/tutorials/user/yaml.rst (hunk 2 above)
Comment (aasivas) [reply docs/source-zh/rst_source/tutorials/user/yaml.rst]: Accepted and corrected both here and English version.

File: rlinf/config.py (hunk 1 above)
Comment (aasivas) [reply rlinf/config.py]: Done.

Comment (aasivas) [reply rlinf/config.py]: Done.

**Review by guozhen1997 — APPROVED** [review]
